cell formula calculation having 2 indirect refences to same cell fails on second reference

<!--
If you are reporting a new issue, make sure that we do not have any duplicates
already open. You can ensure this by searching the issue list for this
repository. If there is a duplicate, please close your issue and add a comment
to the existing issue instead.

Use the commands below to provide key information from your environment:
You do NOT have to include this information if this is a FEATURE REQUEST
-->

**Description**

Here is a unit test to show the problem:

```go
func TestSameRefTwice(t *testing.T) {
	f := NewFile()
	assert.NoError(t, f.SetCellValue("Sheet1", "A1", "VALUE1"))
	assert.NoError(t, f.SetCellFormula("Sheet1", "B1", "=A1"))
	for formula, expected := range map[string]string{
		`=CONCATENATE(A1,"_",A1)`: "VALUE1_VALUE1", // 2 direct references to same cell: OK
		`=CONCATENATE(A1,"_",B1)`: "VALUE1_VALUE1", // first direct, 2nd indirect to same cell: OK
		`=CONCATENATE(B1,"_",B1)`: "VALUE1_VALUE1", // 2 indirect references to same cell fails on 2nd reference
	} {
		assert.NoError(t, f.SetCellFormula("Sheet1", "C1", formula))
		result, err := f.CalcCellValue("Sheet1", "C1")
		assert.NoError(t, err, formula)
		assert.Equal(t, expected, result, formula)
	}
}
```
The first 2 formulas work fine. The third one fails.

<!--
Briefly describe the problem you are having in a few paragraphs.
-->

**Describe the results you received:**
```text
=== RUN   TestSameRefTwice
    calc_test.go:5887: 
        	Error Trace:	D:\isc\code\excelize\calc_test.go:5887
        	Error:      	Not equal: 
        	            	expected: "VALUE1_VALUE1"
        	            	actual  : "VALUE1_"
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1 +1 @@
        	            	-VALUE1_VALUE1
        	            	+VALUE1_
        	Test:       	TestSameRefTwice
        	Messages:   	=CONCATENATE(B1,"_",B1)
--- FAIL: TestSameRefTwice (1255.64s)


Expected :VALUE1_VALUE1
Actual   :VALUE1_
```
**Describe the results you expected:**

Successful unit test

**Output of `go version`:**

```text
=== RUN   TestSameRefTwice
    calc_test.go:5887: 
        	Error Trace:	D:\isc\code\excelize\calc_test.go:5887
        	Error:      	Not equal: 
        	            	expected: "VALUE1_VALUE1"
        	            	actual  : "VALUE1_"
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1 +1 @@
        	            	-VALUE1_VALUE1
        	            	+VALUE1_
        	Test:       	TestSameRefTwice
        	Messages:   	=CONCATENATE(B1,"_",B1)
--- FAIL: TestSameRefTwice (1255.64s)


Expected :VALUE1_VALUE1
Actual   :VALUE1_
```

**Excelize version or commit ID:**

```text
Commit 363fa94
```

**Environment details (OS, Microsoft Excel™ version, physical, etc.):**

Windows 10

--- a/calc.go
+++ b/calc.go
@@ -197,9 +197,10 @@ var (
 // calcContext defines the formula execution context.
 type calcContext struct {
 	sync.Mutex
-	entry           string
-	iterations      map[string]uint
-	iterationsCache map[string]formulaArg
+	entry             string
+	maxCalcIterations uint
+	iterations        map[string]uint
+	iterationsCache   map[string]formulaArg
 }
 
 // cellRef defines the structure of a cell reference.
@@ -775,9 +776,10 @@ func (f *File) CalcCellValue(sheet, cell string, opts ...Options) (result string
 		token        formulaArg
 	)
 	if token, err = f.calcCellValue(&calcContext{
-		entry:           fmt.Sprintf("%s!%s", sheet, cell),
-		iterations:      make(map[string]uint),
-		iterationsCache: make(map[string]formulaArg),
+		entry:             fmt.Sprintf("%s!%s", sheet, cell),
+		maxCalcIterations: getOptions(opts...).MaxCalcIterations,
+		iterations:        make(map[string]uint),
+		iterationsCache:   make(map[string]formulaArg),
 	}, sheet, cell); err != nil {
 		return
 	}
@@ -1529,22 +1531,6 @@ func (f *File) cellResolver(ctx *calcContext, sheet, cell string) (formulaArg, e
 		value string
 		err   error
 	)
-	ref := fmt.Sprintf("%s!%s", sheet, cell)
-	if formula, _ := f.GetCellFormula(sheet, cell); len(formula) != 0 {
-		ctx.Lock()
-		if ctx.entry != ref {
-			if ctx.iterations[ref] <= f.options.MaxCalcIterations {
-				ctx.iterations[ref]++
-				ctx.Unlock()
-				arg, _ = f.calcCellValue(ctx, sheet, cell)
-				ctx.iterationsCache[ref] = arg
-				return arg, nil
-			}
-			ctx.Unlock()
-			return ctx.iterationsCache[ref], nil
-		}
-		ctx.Unlock()
-	}
 	if value, err = f.GetCellValue(sheet, cell, Options{RawCellValue: true}); err != nil {
 		return arg, err
 	}
@@ -1558,8 +1544,25 @@ func (f *File) cellResolver(ctx *calcContext, sheet, cell string) (formulaArg, e
 			return newEmptyFormulaArg(), err
 		}
 		return arg.ToNumber(), err
-	default:
+	case CellTypeInlineString, CellTypeSharedString:
 		return arg, err
+	case CellTypeFormula:
+		ref := fmt.Sprintf("%s!%s", sheet, cell)
+		if ctx.entry != ref {
+			ctx.Lock()
+			if ctx.iterations[ref] <= ctx.maxCalcIterations {
+				ctx.iterations[ref]++
+				ctx.Unlock()
+				arg, _ = f.calcCellValue(ctx, sheet, cell)
+				ctx.iterationsCache[ref] = arg
+				return arg, nil
+			}
+			ctx.Unlock()
+			return ctx.iterationsCache[ref], nil
+		}
+		fallthrough
+	default:
+		return newEmptyFormulaArg(), err
 	}
 }
 
@@ -7753,7 +7756,7 @@ func (fn *formulaFuncs) COUNTBLANK(argsList *list.List) formulaArg {
 	}
 	var count float64
 	for _, cell := range argsList.Front().Value.(formulaArg).ToList() {
-		if cell.Value() == "" {
+		if cell.Type == ArgEmpty {
 			count++
 		}
 	}

--- a/calc.go
+++ b/calc.go
@@ -197,8 +197,9 @@ var (
 // calcContext defines the formula execution context.
 type calcContext struct {
 	sync.Mutex
-	entry      string
-	iterations map[string]uint
+	entry           string
+	iterations      map[string]uint
+	iterationsCache map[string]formulaArg
 }
 
 // cellRef defines the structure of a cell reference.
@@ -774,8 +775,9 @@ func (f *File) CalcCellValue(sheet, cell string, opts ...Options) (result string
 		token        formulaArg
 	)
 	if token, err = f.calcCellValue(&calcContext{
-		entry:      fmt.Sprintf("%s!%s", sheet, cell),
-		iterations: make(map[string]uint),
+		entry:           fmt.Sprintf("%s!%s", sheet, cell),
+		iterations:      make(map[string]uint),
+		iterationsCache: make(map[string]formulaArg),
 	}, sheet, cell); err != nil {
 		return
 	}
@@ -1530,11 +1532,16 @@ func (f *File) cellResolver(ctx *calcContext, sheet, cell string) (formulaArg, e
 	ref := fmt.Sprintf("%s!%s", sheet, cell)
 	if formula, _ := f.GetCellFormula(sheet, cell); len(formula) != 0 {
 		ctx.Lock()
-		if ctx.entry != ref && ctx.iterations[ref] <= f.options.MaxCalcIterations {
-			ctx.iterations[ref]++
+		if ctx.entry != ref {
+			if ctx.iterations[ref] <= f.options.MaxCalcIterations {
+				ctx.iterations[ref]++
+				ctx.Unlock()
+				arg, _ = f.calcCellValue(ctx, sheet, cell)
+				ctx.iterationsCache[ref] = arg
+				return arg, nil
+			}
 			ctx.Unlock()
-			arg, _ = f.calcCellValue(ctx, sheet, cell)
-			return arg, nil
+			return ctx.iterationsCache[ref], nil
 		}
 		ctx.Unlock()
 	}

--- a/calc_test.go
+++ b/calc_test.go
@@ -1023,7 +1023,7 @@ func TestCalcCellValue(t *testing.T) {
 		"=COUNTBLANK(MUNIT(1))": "0",
 		"=COUNTBLANK(1)":        "0",
 		"=COUNTBLANK(B1:C1)":    "1",
-		"=COUNTBLANK(C1)":       "1",
+		"=COUNTBLANK(C1)":       "0",
 		// COUNTIF
 		"=COUNTIF(D1:D9,\"Jan\")":     "4",
 		"=COUNTIF(D1:D9,\"<>Jan\")":   "5",
@@ -5872,17 +5872,18 @@ func TestCalcColRowQRDecomposition(t *testing.T) {
 	assert.False(t, calcColQRDecomposition([][]float64{{0, 0}, {0, 0}}, []float64{0, 0}, 1, 0))
 }
 
-func TestSameRefTwice(t *testing.T) {
+func TestCalcCellResolver(t *testing.T) {
 	f := NewFile()
+	// Test reference a cell multiple times in a formula
 	assert.NoError(t, f.SetCellValue("Sheet1", "A1", "VALUE1"))
-	assert.NoError(t, f.SetCellFormula("Sheet1", "B1", "=A1"))
+	assert.NoError(t, f.SetCellFormula("Sheet1", "A2", "=A1"))
 	for formula, expected := range map[string]string{
-		`=CONCATENATE(A1,"_",A1)`: "VALUE1_VALUE1", // 2 direct references to same cell: OK
-		`=CONCATENATE(A1,"_",B1)`: "VALUE1_VALUE1", // first direct, 2nd indirect to same cell: OK
-		`=CONCATENATE(B1,"_",B1)`: "VALUE1_VALUE1", // 2 indirect references to same cell fails on 2nd reference
+		"=CONCATENATE(A1,\"_\",A1)": "VALUE1_VALUE1",
+		"=CONCATENATE(A1,\"_\",A2)": "VALUE1_VALUE1",
+		"=CONCATENATE(A2,\"_\",A2)": "VALUE1_VALUE1",
 	} {
-		assert.NoError(t, f.SetCellFormula("Sheet1", "C1", formula))
-		result, err := f.CalcCellValue("Sheet1", "C1")
+		assert.NoError(t, f.SetCellFormula("Sheet1", "A3", formula))
+		result, err := f.CalcCellValue("Sheet1", "A3")
 		assert.NoError(t, err, formula)
 		assert.Equal(t, expected, result, formula)
 	}

--- a/calc_test.go
+++ b/calc_test.go
@@ -5871,3 +5871,19 @@ func TestCalcColRowQRDecomposition(t *testing.T) {
 	assert.False(t, calcRowQRDecomposition([][]float64{{0, 0}, {0, 0}}, []float64{0, 0}, 1, 0))
 	assert.False(t, calcColQRDecomposition([][]float64{{0, 0}, {0, 0}}, []float64{0, 0}, 1, 0))
 }
+
+func TestSameRefTwice(t *testing.T) {
+	f := NewFile()
+	assert.NoError(t, f.SetCellValue("Sheet1", "A1", "VALUE1"))
+	assert.NoError(t, f.SetCellFormula("Sheet1", "B1", "=A1"))
+	for formula, expected := range map[string]string{
+		`=CONCATENATE(A1,"_",A1)`: "VALUE1_VALUE1", // 2 direct references to same cell: OK
+		`=CONCATENATE(A1,"_",B1)`: "VALUE1_VALUE1", // first direct, 2nd indirect to same cell: OK
+		`=CONCATENATE(B1,"_",B1)`: "VALUE1_VALUE1", // 2 indirect references to same cell fails on 2nd reference
+	} {
+		assert.NoError(t, f.SetCellFormula("Sheet1", "C1", formula))
+		result, err := f.CalcCellValue("Sheet1", "C1")
+		assert.NoError(t, err, formula)
+		assert.Equal(t, expected, result, formula)
+	}
+}

